rfd/0169-auto-updates-linux-agents.md

```diff
@@ -54,15 +54,57 @@ Whether the updater querying the endpoint is instructed to upgrade (via `agent_a
-A host is eligible to upgrade if the label is present on any of its connected resources.
+Rollouts are specified as interdependent groups of hosts, selected by SSH resource or instance label query.
+A host is eligible to upgrade if the seleciton query returns true.
+Instance labels are a new feature introduced by this RFD that may be used when SSH service is not running or it is undesirable to reuse SSH labels:
+
+```
+teleport:
+  labels:
+    environment: staging
+  commands:
+    # this command will add a label 'arch=x86_64' to an instance
+    - name: arch
+      command: ['/bin/uname', '-p']
+      period: 1h0m0s
+```
+
+Only static and command-based and labels may be used.
 
 At the start of a group rollout, the Teleport proxy marks a desired group of hosts to update in the backend.
-An arbitrarily selected fixed number of hosts (`max_in_flight x total`) are instructed to upgrade at the same time via `/v1/webapi/find`.
+An arbitrary but UUID-deterministic fixed number of hosts (`max_in_flight % x total`) are instructed to upgrade at the same time via `/v1/webapi/find`.
 Additional hosts are instructed to update as earlier updates complete, timeout, or fail, never exceeding `max_in_flight`.
 The group rollout is halted if timeouts or failures exceed their specified thresholds.
 Group rollouts may be retried with `tctl autoupdate run`.
 
+### Scalability
+
+Instance heartbeats will now be cached at both the auth server and the proxy.
+
+All rollout logic is trigger by instance heartbeat backend writes, as changes can only occur on these events.
+The following data related to the rollout are stored in each instance heartbeat:
+- `agent_upgrade_start_time`: timestamp of individual agent's upgrade time
+- `agent_upgrade_group_schedule`: schedule type of group (e.g., critical)
+- `agent_upgrade_group_name`: name of group (e.g., staging)
+- `agent_upgrade_group_start_time`: timestamp of current window start time
+- `agent_upgrade_group_end_time`: timestamp of current window start time
+
+At the start of the window, all queried instance heartbeats are marked with updated values for the `agent_upgrade_group_*` fields.
+Instance heartbeats are included in the current window if all three fields match the window defined in `cluster_maintenance_config`.
+
+On each instance heartbeat write, the auth server looks at instance heartbeats in cache and determines if additional agents should be upgrading.
+If they should, additional instance heartbeats are marked as upgrading by setting `agent_upgrade_start_time` to the current time.
+When `agent_upgrade_start_time` is in the group's window, the proxy serves `agent_auto_upgrade: true` when queried via `/v1/webapi/find`.
+
+To avoid synchronization issues between auth servers, the rollout order is deterministically sorted by UUID.
+Two concurrent writes to different auth servers may temporarily result in fewer upgrading instances than desired, but this should be resolved on the next write.
```

Comment:
Disruption could result in agents changing auth servers in the middle of a rollout. This could push one auth server over its max_in_flight limit.

If the rollout order is deterministic, then the system should always converge to `max_in_flight` agents upgrading at a time using only the state of the cached instance heartbeats.

Reply:
We would rely on sorted node UUIDs (same value that would be provided to `/find`).

Can you provide more detail about your concerns?

Reply:
@rosstimothy How about this: at the beginning of the window, one auth server builds a plan with an arbitrary ordering of UUIDs, and writes the plan to the backend. After the plan is written, we have a fixed ordering of nodes that can be followed.

The plan can be protected with an optimistic lock, so that the first successful write wins. After a plan is written, auth servers will use it to progress the rollout during heartbeat writes.

Reply:
(Updated RFD to use this)